draft-irtf-cfrg-hash-to-curve.md

```diff
@@ -1581,75 +1581,143 @@ Steps:
+corresponding Montgomery curve and birational map.
+We describe how to identify such a curve and map immediately below.
+
+### Birational maps from Montgomery to Twisted Edwards curves {#birational-map}
+
+There are two ways to identify the correct Montgomery curve and
+birational map for use when hashing to a given twisted Edwards curve.
+
+When hashing to a standardized twisted Edwards curve for which a corresponding
+Montgomery form and birational map are also standardized, the standard
+Montgomery form and birational map MUST be used to ensure compatibility
+with existing software.
+Two such standardized curves are the edwards25519 and edwards448 curves,
+which correspond to the Montgomery curves curve25519 and curve448, respectively.
+For both of these curves, {{RFC7748}} lists both the Montgomery and twisted Edwards
+forms and gives the corresponding rational maps.
+
+The birational map for edwards25519 ({{RFC7748}}, Section 4.1)
+uses the constant sqrt\_neg\_486664 = sqrt(-486664) mod 2^255 - 19.
+To ensure compatibility, this constant MUST be chosen such that
+sgn0(sqrt\_neg\_486664) == 1.
+Analogous ambiguities in other standardized birational maps MUST be
+resolved in the same way: for any constant k whose sign is ambiguous,
+k MUST be chosen such that sgn0(k) == 1.
+
+The 4-isogeny map from curve448 to edwards448 ({{RFC7748}}, Section 4.2)
+is unambiguous with respect to sign.
+
+When defining new twisted Edwards curves, a Montgomery equivalent and birational
+map SHOULD be specified, and the sign of the birational map SHOULD be stated
+unambiguously.
+
+When hashing to a twisted Edwards curve that does not have a standardized
+Montgomery form or birational map, the following procedure MUST be
+used to derive them.
 For a twisted Edwards curve given by a * x^2 + y^2 = 1 + d * x^2 * y^2,
-first compute A and B, the parameters of the equivalent Montgomery curve,
-as follows:
+first compute A and B, the parameters of the equivalent Montgomery curve
+y'^2 = x'^3 + A * x'^2 + B * x', as follows:
 
-- A = (a + 2) / 2
+- A = (a + d) / 2
 - B = (a - d)^2 / 16
 
-Next, use A and B as the curve parameters in the Elligator 2 method of
-{{elligator2}} to obtain a point (x', y') on the Montgomery curve.
-Finally, convert (x', y') to a point (x, y) on the target curve.
-Letting B' = 4 / (a - d), compute
+Next, let B' = 4 / (a - d). Then the birational map from the point (x', y')
+on the Montgomery curve to the point (x, y) on the twisted Edwards curve is
+given by
 
 - x = x' / y'
 - y = (B' * x' - 1) / (B' * x' + 1)
 
-This can be done in one inversion using Montgomery's trick {{M87}}:
-invert the product y' * (B' * x' + 1), then multiply by y' to obtain
-1 / (B' * x' + 1), and likewise for 1 / y'.
+For completeness, we give the inverse map in {{birational-map-inverse}}.
+Note that the inverse map is not used when hashing to a twisted Edwards curve.
+
+Note that the above map can be evaluated with just one inversion using
```

Comment:
This paragraph can be omitted or be in another place. Montgomery's trick is a well-known technique. The pseudo-code below already contain the trick.

Reply:
Makes sense.